drivers: serial: Add error code to uart_rx_buf_rsp (unstable API change)

nordic-krch opened the pull request:
API did not cover a case when uart_rx_buf_rsp is called when receiver
is already disabled. It may happen if uart_rx_buf_rsp is called too
late and active transfer is already finished. In that case -EAGAIN is
returned.

Added implementation for nrf_uarte.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrf_uarte: Detect late uart_rx_buf_rsp call

Added detection of uart_rx_buf_rsp call which happend to late when
uart is already disabled.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: Add error code to uart_rx_buf_rsp

API did not cover a case when uart_rx_buf_rsp is called when receiver
is already disabled. It may happen if uart_rx_buf_rsp is called too
late and active transfer is already finished. In that case -EACCES is
returned.

Some implementations already returned that error in that case.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
DNM until other implementations are updated.

carlescufi reviewed (approved):
This looks reasonable to me.

anangl reviewed (commented):
Looks good, except for the error code used. I think we should find some more suitable.

de-nordic commented on include/drivers/uart.h:
```diff
@@ -560,6 +560,7 @@ static inline int z_impl_uart_rx_enable(struct device *dev, uint8_t *buf,
  *
  * @retval -ENOTSUP If not supported.
  * @retval -EBUSY   Next buffer already set.
+ * @retval -EAGAIN  Receiver is already disabled (function called too late?).
```
What is returned from other devices in such case?

de-nordic replied:
If we do not distinguish between disabled and non-existent devices, `NODEV` could be used.or use some of operational errors `ENOTCONN` or `ESHUTDOWN`, which is ` Cannot send after transport endpoint shutdown.` according to linux man page for errno, and quite fits `Receiver is already disabled (function called too late?).`

nordic-krch replied:
`ESHUTDOWN` sounds ok to me, @anangl what do you think?

nordic-krch replied:
turns out that 2/3 implementations already were handling that case and returned `-EACCES`, decided to keep it that way as some may already use it.

nordic-krch committed:
drivers: serial: nrf_uart: Improve detection of late uart_rx_buf_rsp

Removed potential race condition between uart being disabled and
rx_buf_rsp call.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
Removed DNM. Also added a commit for nrf_uart where i added protection against race condition between disabling RX and calling `uart_rx_buf_rsp`.
